src/EventSubscriber/ElasticApmRequestSubscriber.php

```diff
@@ -0,0 +1,238 @@
+
+  /**
+   * The current path.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * Logger service.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * The messenger.
+   *
+   * @var \Drupal\Core\Messenger\MessengerInterface
+   */
+  protected $messenger;
+
+  /**
+   * The actual PHP Agent for the Elastic APM server.
+   *
+   * @var \PhilKra\Agent
+   */
+  protected $phpAgent;
+
+  /**
+   * A flag whether the master request was processed.
+   *
+   * @var bool
+   */
+  protected $processedMasterRequest = FALSE;
+
+  /**
+   * Constructs a ElasticApmRequestSubscriber object.
+   *
+   * @param \Drupal\elastic_apm\ElasticApmInterface $elastic_apm
+   *   The Elastic APM service object.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The logger service.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger.
+   */
+  public function __construct(
+    ElasticApmInterface $elastic_apm,
+    RouteMatchInterface $route_match,
+    LoggerInterface $logger,
+    MessengerInterface $messenger
+  ) {
+    $this->elasticApm = $elastic_apm;
+    $this->routeMatch = $route_match;
+    $this->logger = $logger;
+    $this->messenger = $messenger;
+
+    // Initialize the PHP agent if the Elastic APM config is configured.
+    if ($this->elasticApm->isConfigured()) {
+      $this->phpAgent = $this->elasticApm->getAgent();
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    return [
+      KernelEvents::REQUEST => ['onRequest', 30],
+      KernelEvents::FINISH_REQUEST => ['onFinishRequest', 300],
+    ];
+
+    return $events;
+  }
+
+  /**
+   * Start a transaction for the PHP Agent whenever this event is triggered.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\GetResponseEvent $event
+   *   The request event object.
+   */
+  public function onRequest(GetResponseEvent $event) {
+    // Don't process if Elastic APM is not configured.
+    if (!$this->elasticApm->isConfigured()) {
+      return;
+    }
+
+    // If this is a sub request, only process it if there was no master
+    // request yet. In that case, it is probably a page not found or access
+    // denied page.
+    if ($event->getRequestType() !== HttpKernelInterface::MASTER_REQUEST && $this->processedMasterRequest) {
+      return;
+    }
+
+    // Start a new transaction.
+    try {
+      $transaction = $this->phpAgent->startTransaction($this->routeMatch->getRouteName());
+
+      // Capture database time by wrapping spans around the db queries run.
+      $transaction->setSpans($this->constructDatabaseSpans());
+    }
+    catch (Exception $e) {
+      // Notify the user of the error.
+      $this->messenger->addError(t('An error occurred while trying to send the transaction to the Elastic APM server.'));
```

Comment:
**CHANGE REQUEST**

*Testability*

Line length please + `StringTranslationTrait` i.e. `$this->t()`.